### PR TITLE
feat: add tracing support

### PR DESCRIPTION
**Pull request by stakach**
blocked by https://github.com/wyhaines/opentelemetry-api.cr/issues/2

### CODE REVIEW

File: src/constants.cr [hunk 1]
```diff
@@ -22,6 +22,8 @@ module PlaceOS::Api
   INFLUX_HOST    = ENV["INFLUX_HOST"]?
   INFLUX_ORG     = ENV["INFLUX_ORG"]? || "placeos"
 
+  NEW_RELIC_KEY = ENV["NEW_RELIC_KEY"]?
```
Comment (caspiano) [inline src/constants.cr:25]:
```suggestion
  OPENTELEMETRY_PROVIDER_KEY = ENV["OPENTELEMETRY_PROVIDER_KEY"]?
```

File: src/config.cr [hunk 2]
```diff
@@ -25,4 +28,18 @@ module PlaceOS::Api
     Raven::ActionController::ErrorHandler.new,
     ActionController::LogHandler.new(filters, ms: true)
   )
+
+  if api_key = NEW_RELIC_KEY
```
Comment (caspiano) [inline src/config.cr:32]:
```suggestion
  if telemetry_key = OPENTELEMETRY_PROVIDER_KEY
```

File: src/config.cr [hunk 3]
```diff
@@ -25,4 +28,18 @@ module PlaceOS::Api
...
+    OpenTelemetry.configure do |config|
+      config.service_name = "PlaceOS Rest-API"
+      config.service_version = "1.0.0"
+      config.exporter = OpenTelemetry::Exporter.new(variant: :http) do |exporter|
+        exporter = exporter.as(OpenTelemetry::Exporter::Http)
+        exporter.endpoint = "https://otlp.nr-data.net:4318/v1/traces"
+        headers = HTTP::Headers.new
+        headers["api-key"] = api_key
+        exporter.headers = headers
+      end
```
Comment (caspiano) [inline src/config.cr:42]: I'd rather we put this in a method in `log-backend`

File: src/config.cr [hunk 4]
```diff
@@ -25,4 +28,18 @@ module PlaceOS::Api
...
+  if api_key = NEW_RELIC_KEY
+    OpenTelemetry.configure do |config|
+      config.service_name = "PlaceOS Rest-API"
+      config.service_version = "1.0.0"
+      config.exporter = OpenTelemetry::Exporter.new(variant: :http) do |exporter|
+        exporter = exporter.as(OpenTelemetry::Exporter::Http)
+        exporter.endpoint = "https://otlp.nr-data.net:4318/v1/traces"
+        headers = HTTP::Headers.new
+        headers["api-key"] = api_key
+        exporter.headers = headers
```
Comment (caspiano) [inline src/config.cr:41]:
```suggestion
        exporter.headers = HTTP::Headers{
          "api-key" => api_key
        }
```

File: src/config.cr [hunk 5]
```diff
@@ -25,4 +28,18 @@ module PlaceOS::Api
...
     ActionController::LogHandler.new(filters, ms: true)
   )
+
+  if api_key = NEW_RELIC_KEY
+    OpenTelemetry.configure do |config|
+      config.service_name = "PlaceOS Rest-API"
+      config.service_version = "1.0.0"
+      config.exporter = OpenTelemetry::Exporter.new(variant: :http) do |exporter|
+        exporter = exporter.as(OpenTelemetry::Exporter::Http)
+        exporter.endpoint = "https://otlp.nr-data.net:4318/v1/traces"
```
Comment (caspiano) [inline src/config.cr:38]: Can we make this configurable?

File: src/constants.cr (hunk 1 above)
Comment (caspiano) [reply src/constants.cr:25]: I think we should support

- `OPENTELEMETRY_PROVIDER_KEY`
- `ELASTIC_APM_API_KEY`
- `NEW_RELIC_KEY`

**Review by caspiano — CHANGES_REQUESTED** [review]
I really think we should do this in https://github.com/place-labs/log-backend

It's required by all of our services, and it would be much better if we just added a single require to all of our services than maintain the stanza here like we do for logging at the moment.

**Comment by stakach** [comment]
> I really think we should do this in https://github.com/place-labs/log-backend

yep agreed, can use this pull as a reference

**Comment by caspiano** [comment]
Closing in favour of changes in https://github.com/PlaceOS/rest-api/commit/4afd9c86ca15450faca51861efbda0de39ef158a

**Comment by caspiano** [comment]
Closing in favour of changes in https://github.com/PlaceOS/rest-api/commit/4afd9c86ca15450faca51861efbda0de39ef158a

**Pull request closed without merging**